vnext/Shared/OInstance.cpp

```diff
@@ -458,13 +479,35 @@ void InstanceImpl::loadBundleInternal(std::string &&jsBundleRelativePath, bool s
           m_devSettings->platformName,
           /*dev*/ true,
           /*hot*/ false,
-          m_devSettings->inlineSourceMap);
-
-      // Remote debug executor loads script from a Uri, rather than taking the actual bundle string
-      m_innerInstance->loadScriptFromString(
-          std::make_unique<const JSBigStdString>(m_devSettings->useWebDebugger ? bundleUrl : jsBundleString),
-          bundleUrl,
-          synchronously);
+          m_devSettings->inlineSourceMap,
+          hermesBytecodeVersion);
+
+      // This code is based on the HBC Bundle integration on Android
+      // Ref:
+      // https://github.com/facebook/react-native/blob/24d91268b64c7abbd4b26547ffcc663dc90ec5e7/ReactAndroid/src/main/jni/react/jni/CatalystInstanceImpl.cpp#L231
+      if (isHBCBundle(jsBundleString.c_str())) {
+        auto script = std::make_unique<JSBigStdString>(jsBundleString, false);
+        const char *buffer = script->c_str();
+        uint32_t bufferLength = (uint32_t)script->size();
+        uint32_t offset = 8;
+        while (offset < bufferLength) {
+          uint32_t segment = offset + 4;
+          uint32_t moduleLength = bufferLength < segment ? 0 : *(((uint32_t *)buffer) + offset / 4);
+
+          m_innerInstance->loadScriptFromString(
+              std::make_unique<const JSBigStdString>(std::string(buffer + segment, buffer + moduleLength + segment)),
```

Comment:
could string-view work here too?

Reply:
At first this looks like an ideal situation for string_view usage as the buffer ownership doesn't need to be transferred .. but unfortunately Instance::loadScriptFromString required a instance of JSBigString, which needs to implement (c_str())[https://github.com/facebook/react-native/blob/24d91268b64c7abbd4b26547ffcc663dc90ec5e7/ReactCommon/cxxreact/JSBigString.h#L43] method which should return a null terminated buffer.. which should require a copy which makes std::string a convenient choice here, but yes very inefficient. Fixing this requires making changes to the RN shared code as i understand. Please let me know what you think.

Reply:
> ng into a potentially untrusted buffer,

@asklar I've changed the bigger additions to SafeAdd.. (Though i think we don't need to be super critical about this scenarios as it's devbox only) ..